python/ray/util/multiprocessing/pool.py

```diff
@@ -334,29 +355,52 @@ def __init__(self, pool, func, iterable, chunksize=None):
         self._pool = pool
         self._func = func
         self._next_chunk_index = 0
+        self._finished_iterating = False
         # List of bools indicating if the given chunk is ready or not for all
         # submitted chunks. Ordering mirrors that in the in the ResultThread.
         self._submitted_chunks = []
         self._ready_objects = collections.deque()
-        if not hasattr(iterable, "__len__"):
+        try:
+            self._iterator = iter(iterable)
+        except TypeError:
+            # for compatibility with prior releases, encapsulate non-iterable in a list
             iterable = [iterable]
-        self._iterator = iter(iterable)
-        self._chunksize = chunksize or pool._calculate_chunksize(iterable)
-        self._total_chunks = div_round_up(len(iterable), chunksize)
-        self._result_thread = ResultThread([], total_object_refs=self._total_chunks)
+            self._iterator = iter(iterable)
+        if self._iterator == iterable:
+            # we were passsed an iterator, so do not know the number of samples
```

Comment:
Instead of an if-statement, can we push this logic into `_calculate_chunksize()`? Can we have `_calculate_chunksize()` try calculating `len()`? Then we could catch an error when the iterable doesn't have `len()`, and make the default `chunksize` 1. Would this have any downsides (like the entire iterator being read in `_calculate_chunksize()`)?

If not, can we expand this comment? Something like "Got iterator (which has no `len()` function). Make default `chunksize` 1 instead of using `_calculate_chunksize()`."

Reply:
We could push the logic into `_calculate_chunksize`, I think, without technical side effects. However, I suspect it may be less easy to understand:
* The only other invocation of `_calculate_chunksize` is in `_chunk_and_run`, which guarantees an iterable input. 
* In the current layout, the impact of the argument being an iterable or an iterator on `self._chunksize` and `self._result_list_size` is clear by looking in one place:
```
        if self._iterator == iterable:
            # we were passsed an iterator, so do not know the number of samples
            self._chunksize = chunksize or 1
            result_list_size = 0
        else:
            self._chunksize = chunksize or pool._calculate_chunksize(iterable)
            result_list_size = div_round_up(len(iterable), chunksize)
```

However, if you think it's clearer to merge this into `_calculate_chunksize()` I can certainly do so. If not, I agree with your comment suggestion.

Reply:
Good points– I agree that pushing the logic into `_calculate_chunksize()` may make the code less readable. Let's change the comment then.

Also, do you know if there's a more obvious condition we can use in the if-statement than `self._iterator == iterable`? Something like `isinstance(iterable, Iterable)` or `isinstance(iterable, Iterator)`?

Reply:
Ok, will update the comment. 

And yes, I think I can replace the condition to `isinstance(iterable, Iterator)`. I was worried about iterators that are not explicitly derived from the Iterator class, but (it seems that is not an issue)[https://stackoverflow.com/questions/54737928/how-does-collections-iterable-work-with-isinstance].

I will update and push a commit.